pwiz/data/vendor_readers/ABI/ChromatogramList_ABI.cpp

```diff
@@ -52,6 +52,11 @@ PWIZ_API_DECL ChromatogramList_ABI::ChromatogramList_ABI(const MSData& msd, Wiff
 {
 }
 
+PWIZ_API_DECL ChromatogramList_ABI::~ChromatogramList_ABI()
+{
+    wifffile_->close();

```

Comment:
The *List_ABIs have shared_ptr to a WiffFileImpl (or WiffFile2Impl). It's WiffFile2Impl's destructor that should call close, because it'll be called when both *Lists are destroyed. Calling close manually in the *List destructors is unsafe as the other List may still be using the WiffFile.

Reply:
WiffFile2Impl destructor was calling close, but then this part of VendorReaderTestHarness was preventing the file from being renamed later. I'm not sure how to handle it.

            if (msd_reverse.run.spectrumListPtr.get())
                for (size_t j = 0, end = msd_reverse.run.spectrumListPtr->size(); j < end; ++j)
                    msd_reverse.run.spectrumListPtr->spectrum(end - j - 1);

Reply:
Are you saying that there's a WiffFile2Impl that hasn't been destroyed by the time the rename part of the test happens? Or there's some other way it's preventing the rename? Because I'm not sure how the WiffFile2Impl would survive after msd_reverse goes out of scope (that should destroy both its spectrumListPtr and chromatogramListPtr and thus destroy the shared_ptr<WiffFile2Impl> they hold).

Reply:
I can't figure out why Wiff2Impl isn't being destroyed in that part of the test. There is no issue on the similar line:
 bfs::remove_all(newRawPath); // remove the copy of the RAW file with non-ASCII characters

but without close in the SpectrumList destructor, the line:
 bfs::rename(rawpath, rawpath + ".renamed");

fails. 
There should be no issue with calling close multiple times. It is only closing a SQL connection. I think keeping the close in SpectrumList and ChromatogramList is okay. Another option would be to make a special case in VendorReaderTestHarness similar to:
if (bfs::exists(bfs::path(rawpath) / "Analysis.yep")

Let me know which option you prefer or if you can think of something else.

Reply:
I'll take a look at this.

Reply:
The destructor with its CloseFile call was always getting run as expected, but the CloseFile() call wasn't effective every time for some reason. I was able to fix that by running GC::Collect() before CloseFile(), like this:
```
~WiffFile2Impl()
{
    auto dataReader = DataReader();
    if (dataReader != nullptr)
    {
        System::GC::Collect();
        dataReader->CloseFile(((IList<ISample^>^) allSamples)[0]->Sources[0]);
    }
    System::GC::Collect();
}
```

Can you test whether that works for you as well (and you don't need the close() function anymore)?

Other than that, every time Reader_ABI_Test runs, the WIFF2 API is adding 160MB of debug text to a log file at C:\ProgramData\SCIEX\logs\defaultlog_Reader_ABI_Test.log

Can we disable that? If not I can delete the log file at the beginning of every test to prevent it building up, but it won't let me do it after the test (even in an atexit() function) because the log file is locked.